File: microservice-architecture/circuit-breaker/js/test/test_breaker.js
import { CircuitBreaker } from '../src/breaker.js'

function assertEqual(expected, actual, message) {
  if (expected !== actual) throw new Error(`${message} expected=${expected} actual=${actual}`)
}

const breaker = new CircuitBreaker(2)
assertEqual('closed', breaker.state, 'initial')
breaker.recordFailure()
breaker.recordFailure()
assertEqual('open', breaker.state, 'open')
breaker.probe(true)
assertEqual('closed', breaker.state, 'close')
breaker.recordFailure()
breaker.recordFailure()
breaker.probe(false)
assertEqual('open', breaker.state, 'reopen')
console.log('circuit-breaker(js) tests passed')

File: microservice-architecture/circuit-breaker/python/test/test_breaker.py
import os
import sys
import unittest

sys.path.append(os.getcwd())

from src.breaker import CircuitBreaker  # noqa: E402


class CircuitBreakerTest(unittest.TestCase):
    def test_state_machine(self) -> None:
        breaker = CircuitBreaker(2)
        self.assertEqual("closed", breaker.state)
        breaker.record_failure()
        breaker.record_failure()
        self.assertEqual("open", breaker.state)
        breaker.probe(True)
        self.assertEqual("closed", breaker.state)
        breaker.record_failure()
        breaker.record_failure()
        breaker.probe(False)
        self.assertEqual("open", breaker.state)


if __name__ == "__main__":
    unittest.main()
